module/zfs/zvol.c

```diff
@@ -482,6 +482,22 @@ zvol_replay_write(void *arg1, void *arg2, boolean_t byteswap)
 	return (error);
 }
 
+/*
+ * Replay a TX_CLONE_RANGE ZIL transaction that didn't get committed
+ * after a system failure.
+ */
+static int
+zvol_replay_clone_range(void *arg1, void *arg2, boolean_t byteswap)
+{
+#ifdef TODO
+	zvol_state_t *zv = arg1;
+	lr_clone_range_t *lr = arg2;
+#endif
+
+	/* TODO */
```

Comment:
My thinking was that because it is not possible to clone blocks within ZVOLs, zil_claim() will never bump references. When we implement block cloning for ZVOLs, zvol_replay_clone_range() must be a part of that implementation.

Reply:
Do you plan we add another feature flag after we implement it for ZVOLs?  Otherwise software downgrade may end up in BRT references leaks.

Plus, we already few ZIL records that are implemented for Linux, but not FreeBSD, and they stop ZIL replay early, but I guess not the claim, so unless the replay errors are handled properly alike to frees, it may leak also on OS change.

Reply:
I see what you mean. Yes, we don't want to introduce additional feature flag. I'll try to address this. Thanks.

Reply:
I added check to zil_claim_clone_range() that will deny pool import if we find block cloning entries in the ZVOL's ZIL.

Reply:
If it deny pool import, then it is effectively a hidden feature flag.  If it stops claim at that point -- it may be OK, but not great.  IMO a proper solution, as I have briefly mentioned above, would be to call free for all the records starting from or after one returning replay error.

Reply:
I've implement replay vector for block cloning transaction on ZVOLs by freeing BRT references acquired during zil_claim(). Take a look, please, if this is what you had in mind. Thanks!

Reply:
It is not exactly what I was thinking at this point.  It can be OK for this particular PR. Though returning successful completion without doing anything to data may be not great.

I am trying to understand if there is a possible wider problem, unless I miss how it is solved already.  After log record replay error we stop iteration and we should somehow free all things linked to all following claimed records we are not going to replay, otherwise we may leak not only BRT records, but also claimed earlier indirect TX_WRITE data blocks, etc.  I see zil_replay() calls zil_destroy() at the end, that may end up in zil_destroy_sync(), but I am not getting how it supposed to work in case of error.  I can't look on it deeper today, so, if I am wrong, I'd be happy if somebody would point me where.